### PR TITLE
Add StatisticsEngine and CGAL counterpart

### PR DESCRIPTION
**Pull request by jjerphan**
**Goal**: Introduce mesh statistics by first using CGAL's statistics helpers.

**Context**:

Having mesh statistics is convenient to perform subsequent processing on a mesh. 
Some [CGAL internals](https://github.com/CGAL/cgal/blob/master/Polyhedron/demo/Polyhedron/include/CGAL/statistics_helpers.h) exist to compute mesh statistics but they aren't exposed.

**Main changes**:
 - add `statistics` Python interface (WIP: design TBD)
 - add a `StatisticsEngine` and an implementation for CGAL
 - vendor the CGAL internals to compute statistics

### CODE REVIEW

File: tools/Statistics/CGAL/CGALStatisticsEngine.cpp [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+#include "CGALUtils.h"
+#include "CGALTypes.h"
+#include "statistics_helpers.h"
+
+
+using namespace PyMesh;
+
+void CGALStatisticsEngine::convert_mesh_to_native_format() {
+    Polyhedron polyhedron = CGALUtils::mesh_to_polyhedron<Kernel>(m_vertices, m_faces);
+    CGAL::copy_face_graph(polyhedron, m_mesh);
```
Comment (jjerphan) [inline tools/Statistics/CGAL/CGALStatisticsEngine.cpp]: Now that it works, add an adaptation to directly convert it to `CGAL::Surface_mesh` without using `CGAL::Polyhedron_3`.

**Comment by jjerphan** [comment]
@qnzhou: I need your guidance here regarding the design.

As of now, there is just one interface to CGAL's statistics back-end which simply return a single `MeshStatistics`.

```python
mesh_statistics = pymesh.statistics(vertices, faces)

# statistics stored as attributes
mesh_statistics.min_area
mesh_statistics.max_area
# …
```

We could make it fine-grained and have something like:

```python
min_area, max_area, med_area, mean_area = pymesh.area_statistics(vertices, faces)
```

or even:

```python
min_area, max_area, med_area, mean_area = pymesh.statistics.area(vertices, faces)
```

What would you suggest using here?

**Comment by jjerphan** [comment]
Hello @qnzhou

Have you had time to look at it? :slightly_smiling_face:

File: tools/Statistics/CGAL/CGALTypes.h [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+#include <CGAL/array.h>
+#include <CGAL/Polyhedron_3.h>
+#include <CGAL/Surface_mesh.h>
+#include <CGAL/Simple_cartesian.h>
+#include <CGAL/Nef_polyhedron_2.h>
+#include <CGAL/Point_3.h>
+
+typedef std::array<double, 3>                           Point_type;
+typedef std::array<int, 3>                              Polygon_type;
+typedef CGAL::Simple_cartesian<double>                  Kernel;
```
Comment (sloriot) [inline tools/Statistics/CGAL/CGALTypes.h:14]: This kernel is not safe if you call functions like `does_self_intersect()` as exact predicates are needed to return a certified result. `CGAL::Exact_predicates_inexact_construction_kernel` is a better choice.

File: tools/Statistics/CGAL/CGALTypes.h [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+typedef CGAL::Polyhedron_3<Kernel>                      Polyhedron;
```
Comment (sloriot) [inline tools/Statistics/CGAL/CGALTypes.h:15]: What are you using `Polyhedron_3` for? More precisely, why `Surface_mesh` isn't enough?

File: tools/Statistics/CGAL/CGALTypes.h (hunk 2 above)
Comment (jjerphan) [reply tools/Statistics/CGAL/CGALTypes.h:14]: Good to know, I'll change this.

File: tools/Statistics/CGAL/CGALTypes.h (hunk 3 above)
Comment (jjerphan) [reply tools/Statistics/CGAL/CGALTypes.h:15]: I have been was adapting some code to use (see 879c4e99abd29bd18b02c2c7add12ae04efd49de), and should be able to remove `Polyhedron_3` entirely.

I need to get in this PR again.

**Pull request closed without merging**